### PR TITLE
Use FILE_SHARE_DELETE for log files on Windows.

### LINKED ISSUE #36801: Flaky test: Integration test failure on Windows: TestJSONFileLoggerWithOpts
https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/20312/console

```
21:04:56 --- FAIL: TestJSONFileLoggerWithOpts (0.00s) 21:04:56 	jsonfilelog_test.go:170: open C:\Users\ContainerAdministrator\AppData\Local\Temp\docker-logger-216840801\container.log.1: The process cannot access the file because it is being used by another process. 21:04:56 FAIL
```

cc @johnstep

### LINKED ISSUE #39274: Docker json-file logging driver hangs when rotating files on Windows
**Description**

Docker json-file logging driver hangs when rotating files on Windows. Once a container gets in hung state, no more log entries are saved for it unless docker or the container is restarted.

PS> Get-EventLog -LogName Application -Source Docker shows the following error messages:

```
Failed to log msg "" for logger json-file: error removing oldest log file: remove C:\ProgramData\docker\containers\e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae\e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae-json.log.1: The process cannot access the file because it is being used by another process.
```

(REPEATED UNTIL MITIGATED) 
```
Failed to log msg "" for logger json-file: error closing file: close C:\ProgramData\docker\containers\e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae\e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae-json.log: file already closed.
```

The issue does not happen every time but happens quite frequently.

**Steps to reproduce the issue:**
0. Run a Windows 2019 OS VM with docker using named pipes.
1. Run a docker container on the VM which outputs some log entries. Set this container to use json-file logging driver with 1M file size and 2 file-limit.
2. Create another docker container which is has the docker named pipe mounted. Run periodic ContainerLogs API call for the container in step 1 in this container.
3. Let the setup run until Event log entries start showing the error messages described above. Once this state is reached, no more logs are captured in the json files.

**Describe the results you received:**
1. Container logs stop getting captured on log file rotation.

**Describe the results you expected:**
1. Container logs should not get stuck.

**Additional information you deem important (e.g. issue happens only occasionally):**
This issue does not occur every time. But it happens very frequently when multiple containers are running on multiple VMs.

**Output of `docker version`:**

```
> docker version
Client:
 Version:           18.09.3
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        142dfcedca
 Built:             02/28/2019 06:33:17
 OS/Arch:           windows/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.3
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.10.8
  Git commit:       142dfcedca
  Built:            02/28/2019 06:31:15
  OS/Arch:          windows/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 18.09.3
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd gelf json-file local logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 17763 (17763.1.amd64fre.rs5_release.180914-1434)
Operating System: Windows Server 2019 Datacenter Version 1809 (OS Build 17763.316)
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 15GiB
Name: dc48a85165fcf24
ID: KCKO:GJP5:M7R3:IE45:2LVF:FCYR:MOBJ:DFS7:BBV6:L4XV:DRZY:ZA5M
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Windows 2019 VM on GCP
Mitigations = restart affected container, or restart docker service on Windows.

**docker inspect output on a container which got into this state**
```
> docker inspect e9b
[
    {
        "Id": "e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae",
        "Created": "REDACTED",
        "Path": "REDACTED",
        "Args": [
           "REDACTED"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 3648,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2019-05-28T00:52:02.4274538Z",
            "FinishedAt": "2019-05-28T00:52:00.7740957Z"
        },
        "Image": "sha256:b301392b301239e154b21e03c8b77c03abd14e8ab3459835036a411998b60fd8",
        "ResolvConfPath": "",
        "HostnamePath": "",
        "HostsPath": "",
        "LogPath": "C:\\ProgramData\\docker\\containers\\e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae\\e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae-json.log",
        "Name": "REDACTED",
        "RestartCount": 0,
        "Driver": "windowsfilter",
        "Platform": "windows",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "REDACTED"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {
                    "max-file": "2",
                    "max-size": "1m"
                }
            },
            "NetworkMode": "nat",
            "PortBindings": null,
            "RestartPolicy": {
                "Name": "unless-stopped",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 0,
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "process",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DiskQuota": 0,
            "KernelMemory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": null,
            "PidsLimit": 0,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": null,
            "ReadonlyPaths": null
        },
        "GraphDriver": {
            "Data": {
                "dir": "C:\\ProgramData\\docker\\windowsfilter\\e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae"
            },
            "Name": "windowsfilter"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "REDACTED",
                "Destination": "REDACTED",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
        "Config": {
            "Hostname": "REDACTED",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": [
                "REDACTED",
            ],
            "Image": "REDACTED",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [],
            "OnBuild": null,
            "Labels": {
                "agent.config": "REDACTED",
                "agent.managed": "true"
            },
            "StopSignal": "SIGTERM"
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "e9b7722458801aadcfdc6f37a7d44c077c36d3680a906a305dad878e9cf6a9ae",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "nat": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "93a8f0003991defdec723729ef537a49ecab734ede99e070a8dd1181ff79137e",
                    "EndpointID": "8d8509776404f0d0532d5afe3132c5a4a63e94e0b5fcb94751d53f24e29e1f9a",
                    "Gateway": "REDACTED",
                    "IPAddress": "REDACTED",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "00:15:5d:02:1c:ff",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

### PR DESCRIPTION
**Pull request by cpuguy83**
This (hopefully! 🤞 ) fixes issues where one goroutine tries to delete or rename a file
while another goroutine has the file open (e.g. a log reader).

Closes #39856
Closes #36801
Fixes #39274

### CODE REVIEW

**Comment by thaJeztah** [comment]
@ddebroy @olljanat @vikramhh PTAL

**Comment by olljanat** [comment]
To be honest I don't understand how this would work even on theory?

Sharemode is hardcoded on https://github.com/golang/go/blob/4219aec60ab473fa00f1092034ca801218a5dbe9/src/syscall/syscall_windows.go#L297 and that why there is that very long discussion about changing that hardcoded flag on https://github.com/golang/go/issues/32088

**Comment by cpuguy83** [comment]
@olljanat You're right.
So now I've copied syscall.Open with just the one modification for `FILE_SHARE_DELETE`.
Added a unit test for the functionality.

I had to pull in a few other functions from `os` to mimic `os.OpenFile` just in case there is some edge case... probably not even important.

File: daemon/logger/loggerutils/file_windows.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+import (
+	"os"
+	"syscall"
+	"unsafe"
+
+	"github.com/pkg/errors"
+)
+
+func openFile(name string, flag int, perm os.FileMode) (*os.File, error) {
```
Comment (thaJeztah) [inline daemon/logger/loggerutils/file_windows.go]: Should we move this to pkg/system? In that package we also have some other windows-specific utilities (some of which) can be used as drop-in replacements for their default go implementations, e.g. https://github.com/moby/moby/blob/e554ab558985b686c9c3427275a5e016aa1cdb76/pkg/system/filesys_windows.go#L29-L33

**Comment by olljanat** [comment]
@cpuguy83 Now it it looks likely that it will use `FILE_SHARE_DELETE` share mode but note that it actually only solves a part of the problem. If you remove file which is created with that flag and try open new one without closing original open handle it will still fail. 

Only way to prevent that is rename file for some name which is not used by any other process before removing it. So if I understand this right it should be added to just before `os.Remove` to:
https://github.com/moby/moby/blob/a36dfe758d32e83bdbe037d8e655ab315e1da77b/daemon/logger/loggerutils/logfile.go#L218-L222

**Comment by cpuguy83** [comment]
@oljanat Do you have some example code to reproduce this?
The simple one I came up with (call `openFile` again after the file is removed in the test I added `TestOpenFileDelete` ) does not repro that case.

**Comment by olljanat** [comment]
@cpuguy83 Hmm. It was mentioned on https://github.com/golang/go/issues/32088#issuecomment-502850674 that defaults have been changed on latest versions of Windows and it really looks to be case at least on Windows 10, version 1903 so this simple PowerShell script works just fine on it but fails example on Windows Server 2019:
```powershell
$file = [System.IO.File]::Open("c:\temp\test.txt", "Create", "ReadWrite", "Delete")
Remove-Item -Path c:\temp\test.txt
$file2 = [System.IO.File]::Open("c:\temp\test.txt", "Create", "ReadWrite", "Delete")
```

File: daemon/logger/loggerutils/file_windows.go (hunk 1 above)
Comment (thaJeztah) [reply daemon/logger/loggerutils/file_windows.go]: > I'm intending this as a targeted workaround for a very specific and real issue rather than a generic solution.

Gotcha. I was thinking if other places would also benefit from this, having it in that package would make it a slightly bit more reusable; not a blocker though

> It would be nice if go picked up some change to support this option

Yes, definitely

**Review by tonistiigi — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by andrewhsu** [comment]
I'll see if this helps with master builds that fail on RS1 https://ci.docker.com/public/blue/organizations/jenkins/moby/activity/?branch=master

**Comment by thaJeztah** [comment]
We can cherry-pick once we see master is more stable with this 👍
